### PR TITLE
Betweeness centrality

### PR DESCRIPTION
**Pull request by bmckown**
This is a WIP, but I would really like some feedback.  I used the algorithm found [here] (http://networkx.github.io/documentation/networkx-1.9.1/_modules/networkx/algorithms/centrality/betweenness.html#betweenness_centrality).  The algorithm implemented in that library is not functional in style at all, so I did my best to convert to a more immutable implementation.  However, it seemed like when I did, the code became a little messy and ugly making it harder to follow.  

Any and all comments would be appreciated as to how to clean this up but still maintain good scala practices.

cc @pankajgupta @szymonm

### CODE REVIEW

**Comment by bmckown** [comment]
cc @pankajgupta @szymonm

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package com.twitter.cassovary.algorithms.centrality
+import com.twitter.cassovary.graph._
+import com.twitter.cassovary.graph.tourist.PrevNbrCounter
+
+class BetweenessCentrality(graph: DirectedGraph, normalize: Boolean = true) extends AbstractCentrality(graph) {
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala]: some comments would be good. perhaps a link to the wikipedia article will suffice.

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  (a: Map[A,B], b: Map[A,B])(implicit ev: B => Double) : Map[A,B] = {
+    (a.toList ++ b.toList).groupBy(l => l._1).map{ t => t._1 -> t._2
+      .foldLeft(0.0) { (x,y) => x + y._2 }.asInstanceOf[B] }
+  }
+
+  def _recalc(): Unit = {
+    val tempCentralityValues = graph.foldLeft(new Array[Double](graph.maxNodeId + 1)) {
+      (partialCentralityValues, node) =>
+      val bfs = new BreadthFirstTraverser(graph, GraphDir.OutDir, Seq(node.id), Walk.Limits(), Some(new PrevNbrCounter))
+      val (nodes, depths) = bfs.foldLeft((List.empty[Node], Map.empty[Node, Int])) {
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala]: Not sure this bfs.foldLeft statement is needed as nodes is just bfs.toList

And seems like you need access to all the depths. This is actually present in bfs.depthTracker.infoAllNodes but currently depthTracker is private. I'd suggest adding a 
def allDepths: collection.Map[Int, Int] = depthTracker.infoAllNodes in trait DepthTracker in Traverser.scala and using that.

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  private def combineMap [A, @specialized(Int, Double, Float, Long) B]
+  (a: Map[A,B], b: Map[A,B])(implicit ev: B => Double) : Map[A,B] = {
+    (a.toList ++ b.toList).groupBy(l => l._1).map{ t => t._1 -> t._2
+      .foldLeft(0.0) { (x,y) => x + y._2 }.asInstanceOf[B] }
+  }
+
+  def _recalc(): Unit = {
+    val tempCentralityValues = graph.foldLeft(new Array[Double](graph.maxNodeId + 1)) {
+      (partialCentralityValues, node) =>
+      val bfs = new BreadthFirstTraverser(graph, GraphDir.OutDir, Seq(node.id), Walk.Limits(), Some(new PrevNbrCounter))
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala]: code is not using PrevNbrCounter here, so just leave it out

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ */
+package com.twitter.cassovary.algorithms.centrality
+import com.twitter.cassovary.graph._
+import com.twitter.cassovary.graph.tourist.PrevNbrCounter
+
+class BetweenessCentrality(graph: DirectedGraph, normalize: Boolean = true) extends AbstractCentrality(graph) {
+
+  private def combineMap [A, @specialized(Int, Double, Float, Long) B]
+  (a: Map[A,B], b: Map[A,B])(implicit ev: B => Double) : Map[A,B] = {
+    (a.toList ++ b.toList).groupBy(l => l._1).map{ t => t._1 -> t._2
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala]: Is this basically summing the two values for the same key? It is pretty hard to read.

(a.keys ++ b.keys).map { i => a.getOrElse(0.0) + b.getorElse(0.0) }

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package com.twitter.cassovary.algorithms.centrality
+import com.twitter.cassovary.graph._
+import com.twitter.cassovary.graph.tourist.PrevNbrCounter
+
+class BetweenessCentrality(graph: DirectedGraph, normalize: Boolean = true) extends AbstractCentrality(graph) {
+
+  private def combineMap [A, @specialized(Int, Double, Float, Long) B]
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala]: what is the need to specialize for all these types? Just do what you need below. Also, do you need to parametrize on both A and B (i.e., are you using many key types and many value types)?

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala (hunk 2 above)
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala]: why make a Map of Nodes and not their ids ?

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+  def _recalc(): Unit = {
+    val tempCentralityValues = graph.foldLeft(new Array[Double](graph.maxNodeId + 1)) {
+      (partialCentralityValues, node) =>
+      val bfs = new BreadthFirstTraverser(graph, GraphDir.OutDir, Seq(node.id), Walk.Limits(), Some(new PrevNbrCounter))
+      val (nodes, depths) = bfs.foldLeft((List.empty[Node], Map.empty[Node, Int])) {
+        case ((ns, d), n) =>
+          (ns ++ List(n), d ++ Map(n -> bfs.depth(n.id).get))
+      }
+      val previous = nodes.map { n =>
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/BetweenessCentrality.scala]: The calculations are a bit hard to read/understand. Can you try to explain in semi-english/pseudocode what's going on? What's the role of inboundnodes ("previous) and outboundnodes ("nearestOut") ?

**Comment by pankajgupta** [comment]
Also, from what I can tell it seems like you need some variant of single-source-shortest-path here? I suggest first implementing it to allow you to decompose the betweenness centrality calculation rather than one big lump of code.

**Comment by bmckown** [comment]
Updated a few things.  Still a WIP.  But I would like some feedback to see if y'all think I'm on the right track.  Thanks!  

cc @pankajgupta

**Comment by bmckown** [comment]
Also looks as though I need to merge master onto my branch.  I will do so soon and resolve any merge problems.

**Comment by bmckown** [comment]
Merge conflicts due to some weird formatting issues (thanks intellij!)  Closing this PR and submitting a new clean one.

**Comment by zjpjack** [comment]
This betweeness centrality algorithm not be merged into the latest version?

**Pull request closed without merging**